### ISSUE TITLE
CSLA DataPortalException doesn't track the entire business object when exception is thrown from Business object

### ISSUE THREAD

**Issue by KetanBhojak** [MarimerLLC/csla#2432]
Hi Team,

We have been using CSLA in our Asp.Net Web application since 12-13 years. We are using CSLA 4.3.13.0 in our application since long. Now, recently we decided to upgrade the CSLA Version on our application. We initially used CSLA 5.4.0 and then eventually upgraded to CSLA 5.5.0 and 5.5.1 (which is just released last week). This upgrade activity is still in development phase and we are seeing what all changes we would need in our application to work with latest DLLs and we are also updating things at several places. We have used CSLA DLLs from below URL:

https://www.nuget.org/packages/Csla.AspNet/5.5.1-R21080301

On this, we are facing one issue with DataPortalException not tracking the entire BusinessObject in the ex.BusinessObject property even though the exception was initialized with the BusinessObject. The ex.BusinessObject remains NULL. Please find below details of the same.

CSLA version: 5.5.1.0
OS: Windows
Platform: ASP.Net application

We are throwing DataPortalException from the business layer and catching the same on presentation layer with further exception of capturing the entire Business Object. But we get ex.BusinessObject as NULL on presentation layer. Below are lines of code we have written

Business Object Dataportal_Update or Dataportal_insert method:

`throw new DataPortalException("Message related to exception", this);`

Presentation layer:

```
catch (DataPortalException ex)
{
	if (ex.Message.Contains("abc ex"))
        {
        	Hotel hotelObject = ex.BusinessObject as Hotel; //ex.BusinessObject comes as NULL here. We expect the entire Business Object of Hotel class here.
	}
}

```
ex.BusinessObject was having entire business object when we were using CSLA 4.3.13.0 in past. But in CSLA 5.5.1.0, it comes as NULL.

For your quick reference to replicate same issue, I have created a small test app and re-produced the issue in that. We have kept the logic on Page_Load of default.aspx page only, so the logic will execute on the application startup only. We have attached the zip file of the application here.

Can someone guide us on this and let us know what we can do for this and why this line has stopped working with CSLA 5.5.1.0. Also, if there are any plan to address this in next release of CSLA.
[CSLAUpgrade_TestApp_ExceptionBusinessObjectIssue.zip](https://github.com/MarimerLLC/csla/files/6993410/CSLAUpgrade_TestApp_ExceptionBusinessObjectIssue.zip)

**Comment by rockfordlhotka** [comment]
You mean the business object isn't returned as part of the exception?

The default behavior was changed quite a while ago - probably in CSLA 5.0 - because that value is almost never used by anyone, and it is expensive to serialize the object graph back from the server - especially if it is ignored.

There's a configuration option to turn on the old behavior and have the object graph returned from the server. As always, that object graph is in an indeterminate state. It doesn't get refreshed or "fixed" once the exception occurs, so it could represent half-saved data, that was actually _not saved at all_ due to a database transaction rollback. That sort of thing.

**Comment by KetanBhojak** [comment]
Hello Rocky,

Thank you for your response. And yes, we agree that the particular object graph in our case would be in indeterminate state as the transaction hasn't been completed actually due to exception. But still, we have couple of business cases wherein we are using that business object.

Upon some more research we identified a way around to get the entire business object from the exception. It just needed minor code changes. We modified the code to use InnerException first and then use the BusinessObject from inner exception.

We did below code changes for this:

Business Object Dataportal_Update or Dataportal_insert method:

`throw new DataPortalException("Message related to exception", this);`

Presentation layer:

```
catch (DataPortalException ex)
{
	if (ex.Message.Contains("Message related to exception"))
        {
        	Exception exNew = ex as Exception;
                while (!Object.Equals(null, exNew.InnerException))
                {
                    exNew = exNew.InnerException;
                }
                Hotel hotelObject = ((Csla.DataPortalException)exNew).BusinessObject as Hotel;
	}
}
```

As this way of using BusinessObject from the exception is used only at few places in our application, we are fine doing above code change to have our scenarios working for now. Thanks. :)
